### ISSUE TITLE
NPath Complexity of 315525

### ISSUE THREAD

**Issue by Chowarmaan** [phpmd/phpmd#124]
I have a command line option interface, which formats the output for display to the user when the tool is called wrong.  The options could be values in a config file (XML, INI) or even the command line similar to getopts.  

The function below is being reported as a Cylomatic Complexity of 15, but a NPath Complexity over 300K.

The code will loop through each option and depending on how it is defined (must enter, has a value, value optional, etc..) it will build the usage output.  This class is used to support options from various sources, not just the command line with getopts.

```
public function GetHelpOutput($Title, $Description)
{
    $HelpOutput  = $Title . self::TAB . $Description . self::NEW_LINE;
    $HelpOutput .= 'Available Parameters:' . self::NEW_LINE;

    $LargestWord = 0;
    $IsOptional  = FALSE;
    $IsNeedValue = FALSE;
    foreach ($this->CommandLineOptions_ as $Key => $CommandOption)
    {
        if( ! $CommandOption->Required)
        {
            $IsOptional = TRUE;
        }
        if( $CommandOption->ValueRequired)
        {
            $IsNeedValue = TRUE;
        }
        $Size        = strlen($CommandOption->WordParameter);
        $LargestWord = ($Size > $LargestWord) ? $Size : $LargestWord;
    }

    $MaxBuffer = ($IsOptional ? 4 : 0) + ($IsNeedValue ? 6 : 0) + 4 + $LargestWord;
    foreach ($this->CommandLineOptions_ as $Key => $CommandOption)
    {
        $Parameter  = (strlen($CommandOption->LetterParameter) > 0) ? '-' . $CommandOption->LetterParameter : '  ';
        $Parameter .= ($IsOptional && $CommandOption->Required) ? ' ' : NULL;
        if( strlen($CommandOption->WordParameter) > 0)
        {
            $Parameter .= (substr($Parameter, 0, 1) == '-') ? ' | ' : '   ';
            $Parameter .= '--' . $CommandOption->WordParameter;
        }

        if( $CommandOption->ValueRequired)
        {
            $Parameter .= ' value';
        }

        if( ! $CommandOption->Required )
        {
            $Parameter = '[' . $Parameter . ']';
        }

        $HelpOutput .= str_pad($Parameter, $MaxBuffer, ' ', STR_PAD_RIGHT);
        $HelpOutput .= ' - ' . $CommandOption->Description;
        $HelpOutput .= self::NEW_LINE;
    }
    return $HelpOutput;
}
```

Output from PHPMD 1.5.0

OPTIONS.class:307 The method GetHelpOutput() has a
 Cyclomatic Complexity of 15. The configured cyclomatic complexity threshold is
10.
OPTIONS.class:307 The method GetHelpOutput() has a
n NPath complexity of 315525. The configured NPath complexity threshold is 200.

**Comment by Rarst** [comment]
Related I guess: https://github.com/pdepend/pdepend/issues/158

Hit this recently and it was annoying. It forces to rewrite exactly same logic in a different way.

**Comment by ircmaxell** [comment]
It's not different. It's just an error in the calculation of CC that's resulting in an artificially low number.

**Comment by stefanotorresi** [comment]
@ircmaxell as far as I reckon @manuelpichler is just using "the widely accepted and in all tools used formular to calculate the NPath complexity" and doesn't bother much about the fact that it yields a wrong result.

**Comment by Rarst** [comment]
@ircmaxell low? which do you consider to be the error, lower complexity for `if`? The assumption in discussions around this was error to be higher complexity for ternary.

**Comment by stefanotorresi** [comment]
i guess he meant "high".

**Comment by ircmaxell** [comment]
the fact that ternary gives an artificially higher complexity of 5 vs the correct result of 2 is a bug. It's 5 because of an internal detail of Java. (and yes, I meant high).

**Comment by stefanotorresi** [comment]
@ircmaxell i get that the fact that pdpend is a port of jdpend means the bug was ported as well... but doesn't this make it a fixable implementation detail?

**Comment by manuelpichler** [comment]
The NPath implementation used in pdepend is based on the original sience paper "NPATH: A MEASURE OF EXECUTION PAW COMPLEXITY AND ITS APPLICATIONS", where the formula to calculate the NPath complexity is:

```
NP((exprl))+NP((expr2))+NP((expr3))+2
```

and explains the +2 here:

```
For our purposes, the ? operator can be treated similarly to the if -else statement. The 2 that is included in the NP(?) expression reflects the execution path complexity resulting from this statement (i.e., one path is traversed if (expr 1 ) is True, and another path traversed if (expr 1 ) is False).
```

@ircmaxell can you please explain why this definition should be an internal Java detail?

**Comment by Radio** [comment]
@manuelpichler, The formula is correct but shouldn't the NP((exprX)) give 0 for cases when the expression is just a constant or a variable? So it should still be: 0 + 0 + 0 + 2 = 2 as for if/else ?

**Comment by okdewit** [comment]
Npath complexity is a fairly rough metric when used in practice. Defending the formulaic definition in established scientific papers is pretty much argumentum ad verecundiam: If it doesn't hold up logically in reality then it is bugged from a usability point of view.

Consider this:

```php
public function getValue($bool) {
    $val = 0;
    $bool && ($val=1);
    $bool || ($val=2);
    return $val;
}
```

That's basically an if/else, yet it doesn't add *any* npath complexity. Not that it's a particularly good idea to use boolean operators for if/else decisions, but they *do* work that way in PHP. There are more peculiarities (which no doubt make sense from an "under the hood" point of view): For example, booleans add complexity inside if predicates, but not inside case predicates.

Npath complexity is a reasonably good warning flag, but (in practice) far from an exact number.

**Comment by h0gar** [comment]
> Npath complexity is a reasonably good warning flag, but (in practice) far from an exact number.

The result actually makes no sense at all since it's exponential. You get something like 1024 instead of 32, making it a completely irrelevant metric.

**Comment by dolkensp** [comment]
KumbajaMyLord on Reddit has actually identified the bug that is causing this - and yes, it is a bug, even according to the spec:

https://www.reddit.com/r/programming/comments/42cf4n/on_researching_some_wacky_cyclomatic_complexity/cz9cvgl

**Comment by offsides** [comment]
It's a bug in how PDepend calculates the NPath of the ternary operator: https://github.com/pdepend/pdepend/issues/261

**Comment by kylekatarnls** [comment]
Ternary NPapth calculation has been fixed in PDepend 2.3.0.
